skorch/callbacks/training.py

```diff
@@ -273,3 +278,145 @@ def _sink(self, text, verbose):
+
+    fn : function
+      The function to apply to each parameter separately.
+
+    at : int or callable
+      In case you specify an integer it represents the epoch number the
+      function ``fn`` is applied to the parameters, in case ``at`` is
+      a function it will receive ``net`` as parameter and the function
+      is applied to the parameter once ``at`` returns ``True``.
+
+    schedule : callable or None
+      If specified it determines how ``fn`` and ``at`` are used.
+      This function receives ``net``, ``at`` and ``fn`` and returns
+      the function to be applied on the matched parameters.
+
+    """
+    def __init__(self, patterns, fn=noop, at=1, schedule=None):
+        self.at = at
+        self.fn = fn
+        self.schedule = schedule or self._default_schedule
+        self.patterns = patterns
+
+    def initialize(self):
+        if not isinstance(self.patterns, (list, tuple)):
+            self.patterns = [self.patterns]
+
+        if isinstance(self.at, int):
+            self.at = partial(self._epoch_at, epoch=self.at)
+
+    def named_parameters(self, net):
+        return list(net.module_.named_parameters())
+
+    def filter_parameters(self, patterns, params):
+        for pattern in patterns:
+            pattern_fn = lambda n: fnmatch(n, pattern)
+            pattern_fn = pattern if callable(pattern) else pattern_fn
+            for name, param in params:
+                if pattern_fn(name):
+                    yield name, param
+
+    def _default_schedule(self, net, at, fn):
+        if at(net):
+            return fn
+        return noop
+
+    def _epoch_at(self, net, epoch=1):
+        return len(net.history) == epoch
+
+    def on_epoch_begin(self, net, **kwargs):
+        params = self.named_parameters(net)
+        params = self.filter_parameters(self.patterns, params)
+        map_fn = self.schedule(net, self.at, self.fn)
+
+        for _, p in params:
+            map_fn(p)
+
+
+
+class Freezer(ParamMapper):
+    """Freeze matching parameters in the first epoch. You may
+    specify a specific point in time (either by epoch number or using a callable)
+    when the parameters are frozen using the ``at`` parameter.
+
+    See :class:`.ParamMapper` for details.
+    """
+    def __init__(self, *args, **kwargs):
+        kwargs['at'] = kwargs.get('at', 1)
+        kwargs['fn'] = kwargs.get('fn', freeze_parameter)
+        super().__init__(*args, **kwargs)
+
+
+class Unfreezer(ParamMapper):
+    """Inverse operation of ``Freezer``."""
+    def __init__(self, *args, **kwargs):
+        kwargs['at'] = kwargs.get('at', 1)
+        kwargs['fn'] = kwargs.get('fn', unfreeze_parameter)
+        super().__init__(*args, **kwargs)
+
+
+class Initializer(ParamMapper):
+    """Apply any function on matching parameters in the first epoch.
+
+    Examples
+    --------
+
+    Use ``Initializer`` to initialize all dense layer weights with
+    values sampled from an uniform distribution on epoch begin of
```

Comment:
"to initialize ... values ... on epoch begin of the first epoch" -> "... on the beginning of ..."